src/envoy/auth/jwt.cc

```diff
@@ -0,0 +1,154 @@
+#include "jwt.h"
+
+#include "common/common/base64.h"
+#include "common/common/utility.h"
+#include "openssl/evp.h"
+#include "openssl/rsa.h"
+#include "rapidjson/document.h"
+
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace Envoy {
+namespace Http {
+namespace Auth {
+
+namespace Base64url {
+
+std::string decode(std::string input) {
+  std::replace(input.begin(), input.end(), '-', '+');
+  std::replace(input.begin(), input.end(), '_', '/');
+  switch (input.length() % 4) {
+    case 0:
+      break;
+    case 2:
+      input += "==";
+      break;
+    case 3:
+      input += "=";
+      break;
+    default:
+      /* invalid base64url input*/
+      assert(0);
+  }
+  return Base64::decode(input);
+}
+
+}  // namespace Base64url
+
+namespace Util {
+
+const uint8_t* unsignedCStr(const std::string& str) {
+  return reinterpret_cast<const uint8_t*>(str.c_str());
+}
+
+}  // namespace Util
+
+bssl::UniquePtr<EVP_PKEY> Jwt::evpPkeyFromStr(const std::string& pkey_pem) {
+  std::string pkey_der = Base64::decode(pkey_pem);
+  bssl::UniquePtr<RSA> rsa(RSA_public_key_from_bytes(
+      Util::unsignedCStr(pkey_der), pkey_der.length()));
+  if (rsa == nullptr) {
+    return nullptr;
+  }
+  bssl::UniquePtr<EVP_PKEY> key(EVP_PKEY_new());
+  EVP_PKEY_set1_RSA(key.get(), rsa.get());
+
+  return key;
+}
+
+const EVP_MD* Jwt::evpMdFromAlg(const std::string& alg) {
+  /*
+   * may use
+   * EVP_sha384() if alg == "RS384" and
+   * EVP_sha512() if alg == "RS512"
+   */
+  if (alg == "RS256") {
+    return EVP_sha256();
+  } else {
+    return nullptr;
+  }
+}
+
+bool Jwt::verifySignature(bssl::UniquePtr<EVP_PKEY> key, const std::string& alg,
+                          const uint8_t* signature, size_t signature_len,
+                          const uint8_t* signed_data, size_t signed_data_len) {
+  bssl::UniquePtr<EVP_MD_CTX> md_ctx(EVP_MD_CTX_create());
+  const EVP_MD* md = evpMdFromAlg(alg);
```

Comment:
I tried to do like `bssl::UniquePtr<const EVP_MD> md(...)`, but failed with an error
`
external/boringssl/src/include/openssl/base.h:379:25: error: 'Free' is not a member of 'bssl::internal::DeleterImpl<const env_md_st>'
     DeleterImpl<T>::Free(ptr);
`

How to handle this?